docs/proposals/pod-troubleshooting.md

```diff
@@ -0,0 +1,317 @@
+
+### Automation
+
+Abe is a security engineer tasked with running security audits across all of his
+company's running containers. Even though his company has no standard base
+image, he's able to audit all containers using:
+
+```
+% for pod in $(kubectl get -o name pod); do
+    kubectl exec -m gcr.io/neato/security-audit $pod /security-audit.sh
+  done
+```
+
+## Alternatives Considered
+
+### Ephemeral container
+
+An earlier version of this proposal suggested running an ephemeral container in
+the pod namespaces. The container would not be added to the pod spec and would
+exist only as long as the process it ran. This has the advantage of behaving
+similarly to the current kubectl exec, but it is opaque and likely violates
+design assumptions. We could add constructs to track and report on both
+traditional exec process and exec containers, but this would probably be more
+work than adding to the pod spec. Both are generally useful, and neither
+precludes the other in the future, so we chose mutating the pod spec for
+expedience.
+
+### Attaching Container Type Volume
+
+Combining container volumes ([#831](https://issues.k8s.io/831)) with the ability
+to add volumes to the pod spec would get us most of the way there. One could
+mount a volume of debug utilities at debug time. Docker does not allow adding a
+volume to a running container, however, so this would require a container
+restart. A restart doesn't meet our requirements for troubleshooting.
+
+### Inactive container
+
+If Kubernetes supported the concept of an "inactive" container, we could
+configure it as part of a pod and activate it at debug time. In order to avoid
+coupling the debug tool versions with those of the running containers, we would
+need to ensure the debug image was pulled at debug time. The container could
+then be run with a TTY and attached using kubectl. We would need to figure out a
+solution that allows access the filesystem of other containers.
+
+The downside of this approach is that it requires prior configuration. In
+addition to requiring prior consideration, it would increase boilerplate config.
+A requirement for prior configuration makes it feel like a workaround rather
+than a feature of the platform.
+
+### Implicit Empty Volume
+
+Kubernetes could implicitly create an EmptyDir volume for every pod which would
+then be available as target for either the kubelet or a sidecar to extract a
+package of binaries.
+
+Users would have to be responsible for hosting a package build and distribution
+infrastructure or rely on a public one. The complexity of this solution makes it
+undesirable.
+
+### Standalone Pod in Shared Namespace
+
+Kubernetes could support starting a standalone pod that shares the namespace of
+an existing pod.
+
+This would be a small change to Kubernetes, but it would create edge cases in
+the pod lifecycle that would have to be considered. For example, what happens to
+the debugging pod when the target pod is destroyed?
```

Comment:
Presumably it would be destroyed as well as it is a child of the namespaces. 

I do like this approach because the debug pod may be resource intensive and it gets accounted differently than the debug target.